File: AndroidWebBox/coconut-core/src/main/java/com/sniper/coconut/bridge/BridgeSecurityValidator.kt
package com.sniper.coconut.bridge

import android.webkit.MimeTypeMap
import com.sniper.coconut.utils.Logger

/**
 * Bridge Security Validator
 *
 * Validates bridge calls for security:
 * - Domain whitelist checking
 * - Parameter size limits
 * - Method name validation
 * - Rate limiting
 */
class BridgeSecurityValidator {

    private val tag = "BridgeSecurity"

    // Domain whitelist: empty means all domains allowed
    private val allowedDomains = mutableSetOf<String>()

    // Rate limiting: tracks call count per method
    private val callCounts = HashMap<String, Int>()
    private var lastResetTime = System.currentTimeMillis()

    // Configurable limits
    var maxParamsSize: Int = DEFAULT_MAX_PARAMS_SIZE
    var rateLimitPerMethod: Int = DEFAULT_RATE_LIMIT
    var rateLimitWindowMs: Long = DEFAULT_RATE_LIMIT_WINDOW_MS

    companion object {
        private const val TAG = "BridgeSecurity"

        const val DEFAULT_MAX_PARAMS_SIZE = 1024 * 1024 // 1MB
        const val DEFAULT_RATE_LIMIT = 100 // calls per window
        const val DEFAULT_RATE_LIMIT_WINDOW_MS = 60_000L // 1 minute
    }

    /**
     * Add allowed domain
     */
    fun addAllowedDomain(domain: String) {
        allowedDomains.add(domain)
        Logger.d(tag, "Added allowed domain: $domain")
    }

    /**
     * Set allowed domains list
     */
    fun setAllowedDomains(domains: List<String>) {
        allowedDomains.clear()
        allowedDomains.addAll(domains)
        Logger.d(tag, "Set allowed domains: $domains")
    }

    /**
     * Check if domain whitelist is configured
     */
    fun hasDomainWhitelist(): Boolean = allowedDomains.isNotEmpty()

    /**
     * Validate a domain against the whitelist
     *
     * @param url The full URL to validate
     * @return SecurityResult
     */
    fun validateDomain(url: String): SecurityResult {
        // No whitelist configured = allow all
        if (allowedDomains.isEmpty()) {
            return SecurityResult.Valid
        }

        // Local schemes (file://, coconut://) carry no host and serve app-bundled
        // offline content — exempt from the remote-domain whitelist.
        val scheme = try {
            java.net.URI.create(url).scheme?.lowercase()
        } catch (e: Exception) {
            null
        }
        if (!scheme.isNullOrEmpty() && scheme != "http" && scheme != "https") {
            return SecurityResult.Valid
        }

        val host = extractHost(url)
        if (host.isEmpty()) {
            return SecurityResult.Invalid("Cannot extract host from URL: $url")
        }

        // Check exact match and subdomain match
        val isAllowed = allowedDomains.any { domain ->
            host == domain || host.endsWith(".$domain")
        }

        return if (isAllowed) {
            SecurityResult.Valid
        } else {
            Logger.w(tag, "Domain blocked: $host (allowed: $allowedDomains)")
            SecurityResult.Invalid("Domain not allowed: $host")
        }
    }

    /**
     * Validate parameters size
     */
    fun validateParamsSize(paramsJson: String): SecurityResult {
        if (paramsJson.length > maxParamsSize) {
            Logger.w(tag, "Params too large: ${paramsJson.length} bytes (max: $maxParamsSize)")
            return SecurityResult.Invalid("Params size exceeds limit (${paramsJson.length} > $maxParamsSize)")
        }
        return SecurityResult.Valid
    }

    /**
     * Check rate limit for a method
     */
    fun checkRateLimit(method: String): SecurityResult {
        resetIfNeeded()

        val count = callCounts.getOrDefault(method, 0)
        if (count >= rateLimitPerMethod) {
            Logger.w(tag, "Rate limit exceeded for: $method ($count calls in window)")
            return SecurityResult.Invalid("Rate limit exceeded for method: $method")
        }

        callCounts[method] = count + 1
        return SecurityResult.Valid
    }

    /**
     * Extract host from URL.
     *
     * Uses [java.net.URI.create] instead of [android.webkit.URLUtil] / [android.net.Uri.parse]
     * so this class stays JVM-testable without Robolectric. `android.net.Uri.parse()` returns
     * null on the JVM host (it's a stubbed Android API), which previously made
     * [BridgeSecurityValidator] untestable in pure unit tests.
     *
     * Behavior parity with android.net.Uri.parse for the cases we care about:
     *   - IPv4 / IPv6 / hostname hosts: identical `.host` extraction
     *   - Port stripping: URI.host already returns host without port
     *   - Lowercase normalization: both return lowercase host
     *   - Invalid URL: URI.create throws → catch returns "" (same as Uri.parse returning null)
     */
    private fun extractHost(url: String): String {
        return try {
            val uri = java.net.URI.create(url)
            uri.host ?: ""
        } catch (e: Exception) {
            ""
        }
    }

    /**
     * Reset rate limit counters if window has passed
     */
    private fun resetIfNeeded() {
        val now = System.currentTimeMillis()
        if (now - lastResetTime >= rateLimitWindowMs) {
            callCounts.clear()
            lastResetTime = now
        }
    }
}

/**
 * Security validation result
 */
sealed class SecurityResult {
    object Valid : SecurityResult()
    data class Invalid(val reason: String) : SecurityResult()

    val isValid: Boolean get() = this is Valid
    val message: String get() = if (this is Invalid) reason else ""
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Bridge/BridgeSecurityValidator.swift
import Foundation

public enum SecurityResult {
    case valid
    case invalid(String)

    public var isValid: Bool {
        if case .valid = self { return true }
        return false
    }

    public var message: String {
        if case .invalid(let msg) = self { return msg }
        return ""
    }
}

public class BridgeSecurityValidator: @unchecked Sendable {

    private let tag = "BridgeSecurity"

    // All mutable state guarded by `lock` so the validator is safe to call from
    // any thread. Configuration reads (maxParamsSize etc.) are intentionally
    // also locked because callers may mutate them at runtime.
    private let lock = NSLock()
    private var _allowedDomains: Set<String> = []
    private var _callCounts: [String: Int] = [:]
    private var _lastResetTime: Int64 = Int64(Date().timeIntervalSince1970 * 1000)

    private var _maxParamsSize: Int = 1_048_576 // 1MB
    private var _rateLimitPerMethod: Int = 100
    private var _rateLimitWindowMs: Int64 = 60_000 // 1 minute

    public var maxParamsSize: Int {
        get { lock.lock(); defer { lock.unlock() }; return _maxParamsSize }
        set { lock.lock(); defer { lock.unlock() }; _maxParamsSize = newValue }
    }
    public var rateLimitPerMethod: Int {
        get { lock.lock(); defer { lock.unlock() }; return _rateLimitPerMethod }
        set { lock.lock(); defer { lock.unlock() }; _rateLimitPerMethod = newValue }
    }
    public var rateLimitWindowMs: Int64 {
        get { lock.lock(); defer { lock.unlock() }; return _rateLimitWindowMs }
        set { lock.lock(); defer { lock.unlock() }; _rateLimitWindowMs = newValue }
    }

    public func addAllowedDomain(_ domain: String) {
        lock.lock(); defer { lock.unlock() }
        _allowedDomains.insert(domain)
    }

    public func setAllowedDomains(_ domains: [String]) {
        lock.lock(); defer { lock.unlock() }
        _allowedDomains = Set(domains)
    }

    /// Returns the currently configured allowed domains (snapshot under lock).
    public func getAllowedDomains() -> [String] {
        lock.lock(); defer { lock.unlock() }
        return Array(_allowedDomains)
    }

    public func validateDomain(_ url: String) -> SecurityResult {
        let domains = { () -> Set<String> in
            lock.lock(); defer { lock.unlock() }
            return _allowedDomains
        }()

        if domains.isEmpty { return .valid }

        guard let host = extractHost(url), !host.isEmpty else {
            return .invalid("Cannot extract host from URL: \(url)")
        }

        let isAllowed = domains.contains { domain in
            host == domain || host.hasSuffix(".\(domain)")
        }

        if isAllowed {
            return .valid
        } else {
            return .invalid("Domain not allowed: \(host)")
        }
    }

    public func validateParamsSize(_ json: String) -> SecurityResult {
        let limit = { () -> Int in
            lock.lock(); defer { lock.unlock() }
            return _maxParamsSize
        }()
        if json.utf8.count > limit {
            return .invalid("Params size exceeds limit (\(json.utf8.count) > \(limit))")
        }
        return .valid
    }

    public func checkRateLimit(_ method: String) -> SecurityResult {
        lock.lock(); defer { lock.unlock() }

        resetIfNeededLocked()

        let count = _callCounts[method, default: 0]
        if count >= _rateLimitPerMethod {
            return .invalid("Rate limit exceeded for method: \(method)")
        }

        _callCounts[method] = count + 1
        return .valid
    }

    private func extractHost(_ url: String) -> String? {
        guard let urlObj = URL(string: url) else { return nil }
        return urlObj.host
    }

    /// Must be called while holding `lock`.
    private func resetIfNeededLocked() {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        if now - _lastResetTime >= _rateLimitWindowMs {
            _callCounts.removeAll()
            _lastResetTime = now
        }
    }

    /// Test/support hook: clears rate-limit counters and resets the window.
    public func resetRateLimit() {
        lock.lock(); defer { lock.unlock() }
        _callCounts.removeAll()
        _lastResetTime = Int64(Date().timeIntervalSince1970 * 1000)
    }
}
